x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1137,6 +1148,235 @@ protected Expression rule(Expression e) {
+                opRight = operation.right();
+
+                if (opLeft instanceof Literal) {
+                    opLiteral = (Literal) opLeft;
+                } else if (opRight instanceof Literal) {
+                    opLiteral = (Literal) opRight;
+                } else {
+                    opLiteral = null;
+                }
+            }
+
+            // can it be quickly fast-tracked that the operation can't be reduced?
+            boolean cannotSimplify(BiFunction<DataType, DataType, Boolean> typesCompatible) {
+                if (opLiteral == null) {
+                    // one of the arithm. operands must be a literal, otherwise the operation wouldn't simplify anything
+                    return true;
+                }
+
+                // Only operations on fixed point literals are supported, since optimizing float point operations can also change the
+                // outcome of the filtering:
+                //    x + 1e18 > 1e18::long will yield different results with a field value in [-2^6, 2^6], optimised vs original;
+                //    x * (1 + 1e-15d) > 1 : same with a field value of (1 - 1e-15d)
+                // so consequently, int fields optimisation requiring FP arithmetic isn't possible either: (x - 1e-15) * (1 + 1e-15) > 1.
+                if (opLiteral.dataType().isRational() || bcLiteral.dataType().isRational()) {
+                    return true;
+                }
+
+                // the Literal will be moved to the right of the comparison, but only if data-compatible with what's there
+                return typesCompatible.apply(bcLiteral.dataType(), opLiteral.dataType()) == false;
+            }
+
+            Expression doSimplify() {
+                // force float point folding for FlP field
+                Literal bcl = operation.dataType().isRational()
+                    ? Literal.of(bcLiteral, ((Number) bcLiteral.value()).doubleValue())
+                    : bcLiteral;
+
+                Expression bcRightExpression = safeMaybeFold(operation.inverse(bcl.source(), bcl, opRight));
+                return bcRightExpression == null ? comparison : comparison.replaceChildren(List.of(opLeft, bcRightExpression));
+            }
+
+            static Expression simplify(BinaryComparison comparison, BiFunction<DataType, DataType, Boolean> typesCompatible) {
+                ArithmeticOperation operation = (ArithmeticOperation) comparison.left();
+                // Use symbol comp: SQL operations aren't available in this package (as dependencies)
+                String opSymbol = operation.symbol();
+                // Modulo can't be simplified.
+                if (MOD.symbol().equals(opSymbol)) {
+                    return comparison;
+                }
+                SimplifyOperation simplification = (MUL.symbol().equals(opSymbol) || DIV.symbol().equals(opSymbol))
+                    ? new SimplifyMulDiv(comparison)
+                    : new SimplifyAddSub(comparison);
```

Comment:
Remove the explicit check on the not handled cases (MUL), just to stay on the safe side in case we add any sort of magic Arithmetic operators (so any possible new operation won't be treated as Add/Sub by default).

```suggestion
                SimplifyOperation simplification = null;
                if (isMulOrDiv(comparison.left()) {
                    simplification = new SimplifyMulDiv(comparison);
                } else if (isAddOrSub(comparison.left()) {
                    simplification = new SimpliAddSub(comparison);
                }
                if (simplification == null) {
                    return comparison;
                }
```

Reply:
I was wrongly thinking mostly at SQL (and less at QL) and assuming we won't add operators (without at least breaking these tests), but sure, it absolutely makes sense to expand the check.